### PR TITLE
fix: harden startup and input edge cases

### PR DESCRIPTION
**Pull request by zouyonghe**
## Summary
- Reject unsupported startup files instead of silently opening the parent directory, while preserving stat errno for clearer path errors.
- Harden input and CLI edge cases: strict SGR mouse parsing, unsigned input bytes, extra positional arg rejection, and clearer config/CLI enum errors.
- Preserve video pause state across scale/resize flows, keep file-manager refresh failures from corrupting state, and factor common mode-entry cleanup.
- Organize test link objects in the Makefile and document manual checksum verification.

## Review Follow-up
- Fixed incompatible g_test_queue_destroy callback signature in file-manager tests.
- Removed stale APP_STARTUP_PATH_PARENT_DIRECTORY enum after removing fallback behavior.

## Testing
- rtk make test

Co-Authored-By: Warp <agent@warp.dev>

## Summary by Sourcery

在强化启动阶段的校验、输入处理以及视频/文件管理器行为的同时，改进 CLI/配置/安装相关错误信息，并整理共享的测试与链接基础设施。

Bug 修复：
- 拒绝不支持的启动文件，并返回明确、基于 errno 的错误，而不是回退到父目录。
- 确保文件管理器刷新失败时能保留之前的目录、条目以及选中状态。
- 在尺寸调整和缩放操作中保持视频暂停状态不变，并在视频加载失败时避免静默退回到图片渲染。
- 将格式错误的 SGR 鼠标序列和无效输入字节视为未知输入事件，而不是被误解析为鼠标动作。
- 拒绝多余的 CLI 位置参数，并对 protocol、text symbols、color enhance 和 kitty-transfer 等枚举选项进行校验，明确给出期望值。

功能增强：
- 把共享的模式退出清理逻辑抽取到一个辅助函数中，用于停止媒体、隐藏叠加层，并在需要时清空预加载队列。
- 在加载单个文件时优先使用完整的规范路径匹配，以避免跨目录的同名文件冲突。
- 明确 CLI、配置和安装器的错误信息，包括布尔选项的变体说明以及对不支持的安装标志的使用提示。
- 重构 Makefile 中的测试链接对象，将其整理为可复用的、按组件划分的分组，覆盖 common、render、media、input、app 和 terminal 代码。

文档：
- 记录安装器使用 SHA256SUMS 进行校验的流程，并刷新 README 中的使用示例和固定版本说明。

测试：
- 扩展围绕 CLI 解析、配置枚举、启动分类 errno 传播、文件管理器刷新行为、单文件规范路径匹配、视频尺寸调整/缩放流程、输入边界情况以及安装器参数处理的测试用例。

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Harden startup validation, input handling, and video/file-manager behavior, while improving CLI/config/install error messages and organizing shared test/link infrastructure.

Bug Fixes:
- Reject unsupported startup files with clear errno-backed errors instead of falling back to parent directories.
- Ensure file-manager refresh failures preserve the previous directory, entries, and selection state.
- Keep paused video state intact across resize and scale operations and avoid silently falling back to image rendering on video load failures.
- Treat malformed SGR mouse sequences and invalid input bytes as unknown input events instead of misparsed mouse actions.
- Reject extra CLI positional arguments and validate enum options for protocol, text symbols, color enhance, and kitty-transfer with explicit expected values.

Enhancements:
- Factor shared mode-entry cleanup into a helper that stops media, hides overlays, and optionally clears the preloader queue.
- Prefer full canonical-path matching when loading a single file to avoid basename clashes across directories.
- Clarify CLI, config, and installer error messages, including boolean option variants and guidance for unsupported install flags.
- Refactor Makefile test link objects into reusable component-based groups for common, render, media, input, app, and terminal code.

Documentation:
- Document installer checksum verification using SHA256SUMS and refresh README usage examples and pinned version guidance.

Tests:
- Expand tests around CLI parsing, config enums, startup classification errno propagation, file-manager refresh behavior, single-file canonical matching, video resize/scale flows, input edge cases, and installer argument handling.

</details>

Bug 修复：
- 拒绝不支持的启动文件，而不是回退到父目录，并在路径错误信息中暴露底层的 errno。
- 确保文件管理器刷新失败时保留之前的目录、条目和选中状态，而不是破坏现有状态。
- 在调整终端大小或更改视频缩放时保留已暂停的视频状态，避免视频在这些操作中被意外恢复播放。
- 防止单视图视频加载失败时默默回退到以图像方式渲染。
- 将格式错误的 SGR 鼠标序列和无效输入字节视为未知输入，而不是被误解析为事件。
- 拒绝多余的 CLI 位置参数，并对 protocol、text symbols、color enhance 和 kitty-transfer 选项的枚举值进行有效性校验。

Enhancements（增强功能）：
- 将通用的模式退出清理逻辑提取为一个辅助函数，用于停止媒体、隐藏叠加层，并在需要时清空预加载队列。
- 在加载单个文件时，优先使用完整规范路径匹配，以避免不同目录之间仅按文件名匹配造成的冲突。
- 明确 CLI、配置和安装程序中的错误信息，提供明确的期望值和使用指导，包括布尔选项的变体说明以及安装参数帮助。
- 重构 Makefile 中测试的链接对象，将其按 common、render、media、input、app 和 terminal 组件分组为可共享的集合。

Documentation（文档）：
- 记录安装程序的校验和验证流程，包括手动使用 SHA256SUMS，并刷新 README 中的示例和使用说明。

Tests（测试）：
- 扩展对 CLI、启动流程、文件管理器、输入、视频、安装程序以及单文件加载的测试覆盖，以捕获新增的边缘情况和错误处理行为。

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

在强化启动阶段的校验、输入处理以及视频/文件管理器行为的同时，改进 CLI/配置/安装相关错误信息，并整理共享的测试与链接基础设施。

Bug 修复：
- 拒绝不支持的启动文件，并返回明确、基于 errno 的错误，而不是回退到父目录。
- 确保文件管理器刷新失败时能保留之前的目录、条目以及选中状态。
- 在尺寸调整和缩放操作中保持视频暂停状态不变，并在视频加载失败时避免静默退回到图片渲染。
- 将格式错误的 SGR 鼠标序列和无效输入字节视为未知输入事件，而不是被误解析为鼠标动作。
- 拒绝多余的 CLI 位置参数，并对 protocol、text symbols、color enhance 和 kitty-transfer 等枚举选项进行校验，明确给出期望值。

功能增强：
- 把共享的模式退出清理逻辑抽取到一个辅助函数中，用于停止媒体、隐藏叠加层，并在需要时清空预加载队列。
- 在加载单个文件时优先使用完整的规范路径匹配，以避免跨目录的同名文件冲突。
- 明确 CLI、配置和安装器的错误信息，包括布尔选项的变体说明以及对不支持的安装标志的使用提示。
- 重构 Makefile 中的测试链接对象，将其整理为可复用的、按组件划分的分组，覆盖 common、render、media、input、app 和 terminal 代码。

文档：
- 记录安装器使用 SHA256SUMS 进行校验的流程，并刷新 README 中的使用示例和固定版本说明。

测试：
- 扩展围绕 CLI 解析、配置枚举、启动分类 errno 传播、文件管理器刷新行为、单文件规范路径匹配、视频尺寸调整/缩放流程、输入边界情况以及安装器参数处理的测试用例。

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Harden startup validation, input handling, and video/file-manager behavior, while improving CLI/config/install error messages and organizing shared test/link infrastructure.

Bug Fixes:
- Reject unsupported startup files with clear errno-backed errors instead of falling back to parent directories.
- Ensure file-manager refresh failures preserve the previous directory, entries, and selection state.
- Keep paused video state intact across resize and scale operations and avoid silently falling back to image rendering on video load failures.
- Treat malformed SGR mouse sequences and invalid input bytes as unknown input events instead of misparsed mouse actions.
- Reject extra CLI positional arguments and validate enum options for protocol, text symbols, color enhance, and kitty-transfer with explicit expected values.

Enhancements:
- Factor shared mode-entry cleanup into a helper that stops media, hides overlays, and optionally clears the preloader queue.
- Prefer full canonical-path matching when loading a single file to avoid basename clashes across directories.
- Clarify CLI, config, and installer error messages, including boolean option variants and guidance for unsupported install flags.
- Refactor Makefile test link objects into reusable component-based groups for common, render, media, input, app, and terminal code.

Documentation:
- Document installer checksum verification using SHA256SUMS and refresh README usage examples and pinned version guidance.

Tests:
- Expand tests around CLI parsing, config enums, startup classification errno propagation, file-manager refresh behavior, single-file canonical matching, video resize/scale flows, input edge cases, and installer argument handling.

</details>

</details>

Bug Fixes（缺陷修复）:
- 拒绝不支持的启动文件，而不是回退到其父目录，并保留 errno 以获得更清晰的路径错误信息。
- 确保文件管理器刷新失败时保留原有目录、条目和选中状态不变。
- 在窗口尺寸变化和视频缩放变更时保留暂停视频的状态，避免视频被意外恢复播放。
- 防止单视图视频加载失败时默默回退到图片渲染。
- 稳健处理格式异常的 SGR 鼠标输入和全角按键变体，对有效按键进行映射，并将无效序列视为未知输入。
- 拒绝多余的 CLI 位置参数，并对 protocol、text symbols、color enhance 和 kitty-transfer 选项中的枚举值进行校验，明确报告无效枚举值及其允许取值范围。

Enhancements（增强改进）:
- 将模式切换时的清理逻辑集中到共享辅助函数中，用于停止媒体播放、隐藏信息浮层，并在需要时清空预加载队列。
- 在单文件加载时优先按完整规范路径而非仅按文件名进行匹配，以避免跨目录冲突。
- 优化安装脚本行为，对未知参数给出有用的提示，并优先使用 `--bin-dir` 而不是旧标记。
- 明确 CLI、配置和安装器的错误信息，提供详细预期说明和帮助指引。
- 重构 Makefile 中的测试链接对象分组，在各测试套件中共享通用测试依赖。

Documentation（文档）:
- 记录安装器的校验和验证流程和手动使用 SHA256SUMS 的方法，并更新 README 中的使用说明和版本固定示例。

Tests（测试）:
- 增加以下场景的测试覆盖：启动路径 errno 暴露及非媒体文件拒绝、文件管理器刷新失败行为、单文件加载中的规范路径匹配、视频在尺寸/缩放变化流程中的暂停状态保留、严格的 SGR 鼠标解析、全角波浪号映射、CLI 枚举和布尔值校验、额外位置参数处理、视频加载失败行为，以及安装器参数错误。

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

在强化启动阶段的校验、输入处理以及视频/文件管理器行为的同时，改进 CLI/配置/安装相关错误信息，并整理共享的测试与链接基础设施。

Bug 修复：
- 拒绝不支持的启动文件，并返回明确、基于 errno 的错误，而不是回退到父目录。
- 确保文件管理器刷新失败时能保留之前的目录、条目以及选中状态。
- 在尺寸调整和缩放操作中保持视频暂停状态不变，并在视频加载失败时避免静默退回到图片渲染。
- 将格式错误的 SGR 鼠标序列和无效输入字节视为未知输入事件，而不是被误解析为鼠标动作。
- 拒绝多余的 CLI 位置参数，并对 protocol、text symbols、color enhance 和 kitty-transfer 等枚举选项进行校验，明确给出期望值。

功能增强：
- 把共享的模式退出清理逻辑抽取到一个辅助函数中，用于停止媒体、隐藏叠加层，并在需要时清空预加载队列。
- 在加载单个文件时优先使用完整的规范路径匹配，以避免跨目录的同名文件冲突。
- 明确 CLI、配置和安装器的错误信息，包括布尔选项的变体说明以及对不支持的安装标志的使用提示。
- 重构 Makefile 中的测试链接对象，将其整理为可复用的、按组件划分的分组，覆盖 common、render、media、input、app 和 terminal 代码。

文档：
- 记录安装器使用 SHA256SUMS 进行校验的流程，并刷新 README 中的使用示例和固定版本说明。

测试：
- 扩展围绕 CLI 解析、配置枚举、启动分类 errno 传播、文件管理器刷新行为、单文件规范路径匹配、视频尺寸调整/缩放流程、输入边界情况以及安装器参数处理的测试用例。

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Harden startup validation, input handling, and video/file-manager behavior, while improving CLI/config/install error messages and organizing shared test/link infrastructure.

Bug Fixes:
- Reject unsupported startup files with clear errno-backed errors instead of falling back to parent directories.
- Ensure file-manager refresh failures preserve the previous directory, entries, and selection state.
- Keep paused video state intact across resize and scale operations and avoid silently falling back to image rendering on video load failures.
- Treat malformed SGR mouse sequences and invalid input bytes as unknown input events instead of misparsed mouse actions.
- Reject extra CLI positional arguments and validate enum options for protocol, text symbols, color enhance, and kitty-transfer with explicit expected values.

Enhancements:
- Factor shared mode-entry cleanup into a helper that stops media, hides overlays, and optionally clears the preloader queue.
- Prefer full canonical-path matching when loading a single file to avoid basename clashes across directories.
- Clarify CLI, config, and installer error messages, including boolean option variants and guidance for unsupported install flags.
- Refactor Makefile test link objects into reusable component-based groups for common, render, media, input, app, and terminal code.

Documentation:
- Document installer checksum verification using SHA256SUMS and refresh README usage examples and pinned version guidance.

Tests:
- Expand tests around CLI parsing, config enums, startup classification errno propagation, file-manager refresh behavior, single-file canonical matching, video resize/scale flows, input edge cases, and installer argument handling.

</details>

Bug 修复：
- 拒绝不支持的启动文件，而不是回退到父目录，并在路径错误信息中暴露底层的 errno。
- 确保文件管理器刷新失败时保留之前的目录、条目和选中状态，而不是破坏现有状态。
- 在调整终端大小或更改视频缩放时保留已暂停的视频状态，避免视频在这些操作中被意外恢复播放。
- 防止单视图视频加载失败时默默回退到以图像方式渲染。
- 将格式错误的 SGR 鼠标序列和无效输入字节视为未知输入，而不是被误解析为事件。
- 拒绝多余的 CLI 位置参数，并对 protocol、text symbols、color enhance 和 kitty-transfer 选项的枚举值进行有效性校验。

Enhancements（增强功能）：
- 将通用的模式退出清理逻辑提取为一个辅助函数，用于停止媒体、隐藏叠加层，并在需要时清空预加载队列。
- 在加载单个文件时，优先使用完整规范路径匹配，以避免不同目录之间仅按文件名匹配造成的冲突。
- 明确 CLI、配置和安装程序中的错误信息，提供明确的期望值和使用指导，包括布尔选项的变体说明以及安装参数帮助。
- 重构 Makefile 中测试的链接对象，将其按 common、render、media、input、app 和 terminal 组件分组为可共享的集合。

Documentation（文档）：
- 记录安装程序的校验和验证流程，包括手动使用 SHA256SUMS，并刷新 README 中的示例和使用说明。

Tests（测试）：
- 扩展对 CLI、启动流程、文件管理器、输入、视频、安装程序以及单文件加载的测试覆盖，以捕获新增的边缘情况和错误处理行为。

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

在强化启动阶段的校验、输入处理以及视频/文件管理器行为的同时，改进 CLI/配置/安装相关错误信息，并整理共享的测试与链接基础设施。

Bug 修复：
- 拒绝不支持的启动文件，并返回明确、基于 errno 的错误，而不是回退到父目录。
- 确保文件管理器刷新失败时能保留之前的目录、条目以及选中状态。
- 在尺寸调整和缩放操作中保持视频暂停状态不变，并在视频加载失败时避免静默退回到图片渲染。
- 将格式错误的 SGR 鼠标序列和无效输入字节视为未知输入事件，而不是被误解析为鼠标动作。
- 拒绝多余的 CLI 位置参数，并对 protocol、text symbols、color enhance 和 kitty-transfer 等枚举选项进行校验，明确给出期望值。

功能增强：
- 把共享的模式退出清理逻辑抽取到一个辅助函数中，用于停止媒体、隐藏叠加层，并在需要时清空预加载队列。
- 在加载单个文件时优先使用完整的规范路径匹配，以避免跨目录的同名文件冲突。
- 明确 CLI、配置和安装器的错误信息，包括布尔选项的变体说明以及对不支持的安装标志的使用提示。
- 重构 Makefile 中的测试链接对象，将其整理为可复用的、按组件划分的分组，覆盖 common、render、media、input、app 和 terminal 代码。

文档：
- 记录安装器使用 SHA256SUMS 进行校验的流程，并刷新 README 中的使用示例和固定版本说明。

测试：
- 扩展围绕 CLI 解析、配置枚举、启动分类 errno 传播、文件管理器刷新行为、单文件规范路径匹配、视频尺寸调整/缩放流程、输入边界情况以及安装器参数处理的测试用例。

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Harden startup validation, input handling, and video/file-manager behavior, while improving CLI/config/install error messages and organizing shared test/link infrastructure.

Bug Fixes:
- Reject unsupported startup files with clear errno-backed errors instead of falling back to parent directories.
- Ensure file-manager refresh failures preserve the previous directory, entries, and selection state.
- Keep paused video state intact across resize and scale operations and avoid silently falling back to image rendering on video load failures.
- Treat malformed SGR mouse sequences and invalid input bytes as unknown input events instead of misparsed mouse actions.
- Reject extra CLI positional arguments and validate enum options for protocol, text symbols, color enhance, and kitty-transfer with explicit expected values.

Enhancements:
- Factor shared mode-entry cleanup into a helper that stops media, hides overlays, and optionally clears the preloader queue.
- Prefer full canonical-path matching when loading a single file to avoid basename clashes across directories.
- Clarify CLI, config, and installer error messages, including boolean option variants and guidance for unsupported install flags.
- Refactor Makefile test link objects into reusable component-based groups for common, render, media, input, app, and terminal code.

Documentation:
- Document installer checksum verification using SHA256SUMS and refresh README usage examples and pinned version guidance.

Tests:
- Expand tests around CLI parsing, config enums, startup classification errno propagation, file-manager refresh behavior, single-file canonical matching, video resize/scale flows, input edge cases, and installer argument handling.

</details>

</details>

</details>

---
## Summary by cubic
Reject unsupported startup files and harden input/CLI handling to make errors clearer and behavior safer. Videos keep their play/pause state across scale changes and repeated resize bursts, and file‑manager refresh errors no longer corrupt state.

- **Bug Fixes**
  - Startup: reject non-media files; include `stat` errno in path errors.
  - Input: read bytes as 0–255; strict SGR mouse parsing (require `<`, strip `M/m`, validate ints); malformed or non‑SGR sequences => `KEY_UNKNOWN`.
  - CLI/config/installer: enum errors list expected values (with group names); reject extra positional args; clearer boolean messages; `install.sh` unknown args point to help and suggest `--bin-dir`.
  - Video: preserve paused/playing state across terminal resize bursts and on scale changes; no fallback to image if video load fails.
  - File manager/core: refresh failure keeps previous directory, entries, and selection; single-file load matches canonical path before basename.

- **Refactors**
  - Added `app_prepare_mode_entry()` for shared mode-entry cleanup; grouped test link objects in the Makefile and added targeted regression tests.

<sup>Written for commit a2bb9d89164edf2e8abb5b8529df8f27d30e507e. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/zouyonghe/PixelTerm-C/pull/30?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/input.c [hunk 1]
```diff
@@ -329,9 +349,16 @@ ErrorCode input_get_event(InputHandler *handler, InputEvent *event) {
...
+                        gint button = 0;
+                        gint x = 0;
+                        gint y = 0;
+                        if (!input_parse_sgr_int(btn_str, 0, 255, &button) ||
+                            !input_parse_sgr_int(params[1], 1, G_MAXINT, &x) ||
+                            !input_parse_sgr_int(params[2], 1, G_MAXINT, &y)) {
+                            event->type = INPUT_KEY_PRESS;
+                            event->key_code = KEY_UNKNOWN;
+                            goto finish_event;
+                        }
```
[the remark this answers is not in the document]
Comment (zouyonghe) [reply src/input.c:373]: Fixed across follow-up commits. 39c3b39 strips the SGR M/m terminator before strict Y parsing and adds `/input/parses_sgr_mouse_press_with_terminator`; malformed overflow input remains covered.

File: src/app_core.c [hunk 2]
```diff
@@ -144,20 +144,38 @@ ErrorCode app_load_single_file(PixelTermApp *app, const char *filepath) {
...
+        g_free(current_canonical);
+
+        if (full_path_match) {
             app->current_index = idx;
             found = TRUE;
-            g_free(current_basename);
             break;
         }
-        g_free(current_basename);
     }
```
[the remark this answers is not in the document]
Comment (zouyonghe) [reply src/app_core.c:171]: Addressed in 39c3b39. `app_load_single_file()` now compares the stored path directly to the target canonical path first, then only canonicalizes candidates whose basename matches.

File: src/app_file_manager.c [hunk 3]
```diff
@@ -719,18 +705,32 @@ ErrorCode app_file_manager_refresh(PixelTermApp *app) {
...
-    app->file_manager.directory = current_dir;
 
     // Open directory
     GDir *dir = g_dir_open(current_dir, 0, NULL);
     if (!dir) {
+        if (previous_directory) {
+            g_free(app->file_manager.directory);
+            app->file_manager.directory = g_strdup(previous_directory);
+        }
+        g_free(previous_directory);
```
[the remark this answers is not in the document]
Comment (zouyonghe) [reply src/app_file_manager.c:717]: Addressed in 39c3b39. The refresh failure path now transfers `previous_directory` ownership back to `app->file_manager.directory` and nulls the local pointer before cleanup.

File: tests/test_app_cli.c [hunk 4]
```diff
@@ -611,11 +611,11 @@ static void test_cli_invalid_boolean_values_return_error(AppCliFixture *fixture,
...
                             ==,
-                            "Invalid --preload value: maybe (expected true/false)\n");
+                            "Invalid --preload value: maybe (expected true/false, yes/no, on/off, or 1/0)\n");
         } else {
             g_assert_cmpstr(stderr_output,
                             ==,
-                            "Invalid --alt-screen value: 2 (expected true/false)\n");
+                            "Invalid --alt-screen value: 2 (expected true/false, yes/no, on/off, or 1/0)\n");
         }
         g_free(stderr_output);
```
[the remark this answers is not in the document]
Comment (zouyonghe) [reply tests/test_app_cli.c:620]: Addressed in 39c3b39. Added exact stderr coverage for invalid `--protocol foo` and `--text-symbols ascii` values, including the allowed-value lists.

**Comment by zouyonghe** [comment]
@sourcery-ai review

**Comment by zouyonghe** [comment]
@cubic-dev-ai review

**Comment by zouyonghe** [comment]
@cubic-dev-ai review
